modtools/intro-loader.lua

```diff
@@ -0,0 +1,160 @@
+      - Further, for the -rand argument, usable numbers start at 1 and all
+	    numbers must be consecutive.
+      - The files must not have file extensions
+]====]
+
+
+local utils = require 'utils'
+
+
+validArgs = validArgs or utils.invert({
+    'help',
+    'v'
+    'now',
+    'folder',
+    'file',
+    'rand',
+    'entity',
+})
+
+local args = utils.processArgs({...}, validArgs)
+
+if args.help then
+    print(help)
+    return
+end
+
+local inputfolder = dfhack.getDFPath() .. "\\data\\announcement\\"
+if args.folder then
+    inputfolder = dfhack.getDFPath() ..  args.folder .. "\\"
+end
+
+local prefix = "INTRO"
+if args.file then
+    prefix = args.file
+end
+
+if args.rand then
+    math.randomseed(os.time())
+end
+
+
+--Build current index of files in directory
+do
+    local OS = dfhack.getOSType()
+    if OS == 'windows' then
+        os.execute("dir \"" .. inputfolder .. "\" /L /B >index.txt")
+    elseif OS == 'linux' or OS == 'darwin' then --apple's real name is darwin, apparently
+        os.execute("ls \"" .. inputfolder .. "\" -1 -A >index.txt")
+    end
+    
+    local list = io.open(dfhack.getDFPath() .. "\\data\\announcement\\index.txt","r")
+    indextext = list:read("*all")
+    io.close(list)
+end
+
+
+--Code main body, this is where the magic happens
+local function introloader(folder,dorand,doent,verbose)
+    local entity = "_" .. df.historical_entity.find(df.global.ui.civ_id).entity_raw.code
+    if ~doent then entity = "" end
+    local inputfile = folder .. "\\" .. prefix .. entity
+    
+    if dorand then
+        --Determine number of usable files
+        local filecount = 0
+        while string.gmatch(indextext,inputfile .. "_%d+") do
+            filecount++
+        end
+        
+        --Random decision
+        if filecount > 0 then
+            local rand = math.random(1,filecount)
+            inputfile = inputfile .. "_" .. rand
+        end
+    end
+    
+    if verbose then print("Loading intro text") end
+    local input = assert(io.open(inputfile, "rb"),"No matching intro file, leaving as-is")
+    if input == false then return end
+    local output = assert(io.open(dfhack.getDFPath() .. "\\data\\announcement\\fortressintro", "wb"),"Could not create or open data/announcement/fortressintro, somehow.")
+    if output == false then input:close() return end
+    local t = input:read("*all")
+    t = string.gsub(t, "\r\n", "\n")
+    output:write(t)
+    if verbose then print("Civilization intro text written") end
+    output:close()
+    input:close()
+end
+
+
+
+--Flow control for the code; runs process code after embark selection has been made, to ensure proper operation.
+dfhack.onStateChange.loadNewFortressIntro = function(code)
+if(dorand|doent)
+    if args.v then print("Intro loader ready") end
+    if (code == SC_VIEWSCREEN_CHANGED and dfhack.gui.getCurFocus() == 'setupdwarfgame') or (args.now) then
+            dfhack.with_suspend(introloader,inputfolder,args.rand,args.entity, args.v)
```

Comment:
Yes, because of the file IO being just a bit too slow. I was doing it several times to be sure, and made sure this was specifically the cause, and this was with an earlier, faster version, but essentially...
You move from the map screen to the embark prep screen, which starts the critical section that is the function. If you _immediately_ press enter, you will cause DF to experience an error due to this function being in the middle of writing to fortressintro while DF wants to read from it.
Pausing execution for the incredibly short time here ensures that you _can't_ move to fast, and the normal runtime for this is incredibly short. I almost didn't even notice this error in my earlier testbuilds, but when I started running through my test embarks a bit faster, it became quite obvious that the file I/O was dragging on just a bit too long for DF to deal with.

Additionally the function cannot be effectively run any earlier in DF's execution because your entity hasn't been set until you are leaving the map screen. If I _only_ implemented the random choice loader, it'd be fine running earlier, but the entity-specific code definitely has to be run here, and it's definitely possible for someone with a decent computer to pass through fast enough to cause an error. I can go get you a screencap of what happens if I purposefully set off the error, if you'd like.